### PR TITLE
BUG: support for "level=" when reset_index() is called with a flat Index

### LINKED ISSUE #16263: DataFrame.reset_index deletes index, does not all for ints as level arg
#### Code Sample, a copy-pastable example if possible
```python
import pandas as pd
import numpy as np
pd.__version__
```
  u'0.20.1'

```python
data = pd.DataFrame(np.random.randn(4,4), columns=['A', 'B', 'C', 'D']) 
data
```

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A</th>
      <th>B</th>
      <th>C</th>
      <th>D</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>-0.134549</td>
      <td>2.352525</td>
      <td>0.481132</td>
      <td>-1.919506</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1.980074</td>
      <td>0.720437</td>
      <td>0.410702</td>
      <td>-0.703470</td>
    </tr>
    <tr>
      <th>2</th>
      <td>-3.063166</td>
      <td>-0.781255</td>
      <td>0.270469</td>
      <td>-0.539081</td>
    </tr>
    <tr>
      <th>3</th>
      <td>-1.125265</td>
      <td>0.308374</td>
      <td>-0.166085</td>
      <td>-1.253959</td>
    </tr>
  </tbody>
</table>
</div>

```python
data.set_index(['A'], inplace=True)
data
```
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>B</th>
      <th>C</th>
      <th>D</th>
    </tr>
    <tr>
      <th>A</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>-0.134549</th>
      <td>2.352525</td>
      <td>0.481132</td>
      <td>-1.919506</td>
    </tr>
    <tr>
      <th>1.980074</th>
      <td>0.720437</td>
      <td>0.410702</td>
      <td>-0.703470</td>
    </tr>
    <tr>
      <th>-3.063166</th>
      <td>-0.781255</td>
      <td>0.270469</td>
      <td>-0.539081</td>
    </tr>
    <tr>
      <th>-1.125265</th>
      <td>0.308374</td>
      <td>-0.166085</td>
      <td>-1.253959</td>
    </tr>
  </tbody>
</table>
</div>

```python
data.reset_index(level=['A'], inplace=True)
data
```
<div>

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>B</th>
      <th>C</th>
      <th>D</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2.352525</td>
      <td>0.481132</td>
      <td>-1.919506</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.720437</td>
      <td>0.410702</td>
      <td>-0.703470</td>
    </tr>
    <tr>
      <th>2</th>
      <td>-0.781255</td>
      <td>0.270469</td>
      <td>-0.539081</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.308374</td>
      <td>-0.166085</td>
      <td>-1.253959</td>
    </tr>
  </tbody>
</table>
</div>

#### Problem description

between v0.19.2 and v0.20.1, the behavior of DataFrame.reset_index changed. 
With a single set index:
* It does not attempt to keep the column (essentially making drop=True always on)
* level=int no longer works (iterables work)

#### Expected Output

v0.19.2 results:

```python
data.reset_index(level=['A'], inplace=True)
data
```

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A</th>
      <th>B</th>
      <th>C</th>
      <th>D</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.100442</td>
      <td>-0.620740</td>
      <td>-2.018020</td>
      <td>1.059871</td>
    </tr>
    <tr>
      <th>1</th>
      <td>-0.530272</td>
      <td>0.402598</td>
      <td>-1.453445</td>
      <td>-0.729623</td>
    </tr>
    <tr>
      <th>2</th>
      <td>-1.040126</td>
      <td>-0.536687</td>
      <td>-1.136123</td>
      <td>-0.748891</td>
    </tr>
    <tr>
      <th>3</th>
      <td>-0.269727</td>
      <td>0.182250</td>
      <td>0.847344</td>
      <td>0.785692</td>
    </tr>
  </tbody>
</table>
</div>

#### Output of ``pd.show_versions()``

<details>

```python
import pandas as pd
import numpy as np
pd.show_versions()
```

    INSTALLED VERSIONS
    ------------------
    commit: None
    python: 2.7.13.final.0
    python-bits: 64
    OS: Darwin
    OS-release: 15.6.0
    machine: x86_64
    processor: i386
    byteorder: little
    LC_ALL: None
    LANG: en_US.UTF-8
    LOCALE: None.None
    
    pandas: 0.19.2
    nose: None
    pip: 9.0.1
    setuptools: 35.0.2
    Cython: None
    numpy: 1.12.1
    scipy: None
    statsmodels: None
    xarray: None
    IPython: 5.3.0
    sphinx: None
    patsy: None
    dateutil: 2.6.0
    pytz: 2017.2
    blosc: None
    bottleneck: None
    tables: None
    numexpr: None
    matplotlib: None
    openpyxl: None
    xlrd: None
    xlwt: None
    xlsxwriter: None
    lxml: None
    bs4: None
    html5lib: 0.999999999
    httplib2: None
    apiclient: None
    sqlalchemy: None
    pymysql: None
    psycopg2: None
    jinja2: 2.9.6
    boto: None
    pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #16263
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -77,6 +77,7 @@ Conversion
 Indexing
 ^^^^^^^^
 
+- Bug in ``DataFrame.reset_index(level=)`` with flat index (:issue:`16263`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.21.0.txt]: Can you move this to the 0.20.2 file?

File: pandas/tests/frame/test_alter_axes.py [hunk 2]
```diff
@@ -641,6 +641,36 @@ def test_reset_index(self):
         xp = xp.set_index(['B'], append=True)
         assert_frame_equal(rs, xp, check_names=False)
 
+    def test_reset_index_level(self):
+        df = pd.DataFrame([[1, 2, 3, 4], [5, 6, 7, 8]],
+                          columns=['A', 'B', 'C', 'D'])
+
+        # With MultiIndex
+        for levels in ['A', 'B', 'E'], [0, 1, 2]:
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_alter_axes.py]: Can you remove the 'E' and 2 from the lists? I founded it a bit confusing as you don't use it

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good

File: pandas/tests/frame/test_alter_axes.py (hunk 2 above)
Comment (toobaz) [inline pandas/tests/frame/test_alter_axes.py]: yep, they were refactoring leftovers

File: doc/source/whatsnew/v0.20.2.txt [hunk 3]
```diff
@@ -42,7 +42,7 @@ Conversion
 Indexing
 ^^^^^^^^
 
-
+- Bug in ``DataFrame.reset_index(level=)`` with flat index (:issue:`16263`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.2.txt]: single level index

File: pandas/tests/frame/test_alter_axes.py [hunk 4]
```diff
@@ -641,6 +641,35 @@ def test_reset_index(self):
         xp = xp.set_index(['B'], append=True)
         assert_frame_equal(rs, xp, check_names=False)
 
+    def test_reset_index_level(self):
+        df = pd.DataFrame([[1, 2, 3, 4], [5, 6, 7, 8]],
+                          columns=['A', 'B', 'C', 'D'])
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: issue number

File: pandas/tests/frame/test_alter_axes.py [hunk 5]
```diff
@@ -641,6 +641,35 @@ def test_reset_index(self):
...
+
+        for levels in ['A', 'B'], [0, 1]:
+            # With MultiIndex
+            result = df.set_index(['A', 'B']).reset_index(level=levels[0])
+            tm.assert_frame_equal(result, df.set_index('B'))
+
+            result = df.set_index(['A', 'B']).reset_index(level=levels[:1])
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: can u add similar tests for Series (in series test hierarchy)

File: pandas/tests/frame/test_alter_axes.py [hunk 6]
```diff
@@ -641,6 +641,35 @@ def test_reset_index(self):
...
+            result = df.set_index(['A', 'B']).reset_index(level=levels)
+            tm.assert_frame_equal(result, df)
+
+            # With flat Index (GH 16263)
+            result = df.set_index('A').reset_index(level=levels[0])
+            tm.assert_frame_equal(result, df)
+
+            result = df.set_index('A').reset_index(level=levels[:1])
+            tm.assert_frame_equal(result, df)
+
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: add with drop as well

File: pandas/tests/frame/test_alter_axes.py [hunk 7]
```diff
@@ -641,6 +641,35 @@ def test_reset_index(self):
...
+        # Missing levels - for both MultiIndex and flat Index:
+        for idx_lev in ['A', 'B'], ['A']:
+            with tm.assert_raises_regex(KeyError, 'Level E '):
+                df.set_index(idx_lev).reset_index(level=['A', 'E'])
+            with tm.assert_raises_regex(IndexError, 'Too many levels'):
+                df.set_index(idx_lev).reset_index(level=[0, 1, 2])
```
Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: it might be easier to parameteize these tests (snd make the errors a separate test)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_alter_axes.py (hunk 4 above)
Comment (toobaz) [inline pandas/tests/frame/test_alter_axes.py]: the number is few lines after, it only applies to the flat index

File: pandas/tests/frame/test_alter_axes.py (hunk 5 above)
Comment (toobaz) [inline pandas/tests/frame/test_alter_axes.py]: done

File: pandas/tests/frame/test_alter_axes.py (hunk 6 above)
Comment (toobaz) [inline pandas/tests/frame/test_alter_axes.py]: done

File: pandas/tests/frame/test_alter_axes.py (hunk 7 above)
Comment (toobaz) [inline pandas/tests/frame/test_alter_axes.py]: can you clarify?

Comment (jreback) [inline pandas/tests/frame/test_alter_axes.py]: I guess this is fine.

File: pandas/tests/series/test_alter_axes.py [hunk 8]
```diff
@@ -141,6 +141,45 @@ def test_reset_index(self):
...
+            tm.assert_frame_equal(result, df)
+
+            result = df.set_index(['A', 'B']).reset_index(level=levels,
+                                                          drop=True)
+            tm.assert_frame_equal(result, df[['C']])
+
+            with tm.assert_raises_regex(KeyError, 'Level E '):
+                s.reset_index(level=['A', 'E'])
+
+            # With flat Index
```
Comment (jreback) [inline pandas/tests/series/test_alter_axes.py]: don't use word flat, rather single level

**Comment by jreback** [comment]
minor comment, otherwise lgtm.

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**